### DISCUSSION TITLE
best way for client to get server version

### DISCUSSION THREAD

**Discussion by dfgordon** [microsoft/vscode-discussions#1349] [Extension Development QnA]
I would like my client that uses `vscode-languageclient/node` to check the server version.  I found that the client populates `client.initializeResult.serverInfo.version` at some point during startup, but I can't figure out which event to listen for.  One workaround I found is to do it as a side effect of some unrelated notification.

The question is whether `vscode-languageclient/node` provides something explicitly for this purpose, like something to listen for that fires when the server sends back the `initializeResult`.

I have control of the server and could generate a notification, but in general one might not have that.

**Comment by theReynald** [comment]
Hi Dfgordon.

I interacted with copilot regarding your question and it provided the following suggestion. Let me know if this helps.

**_Copilot Suggestion_**

To check the server version using vscode-languageclient/node, you can listen for the onReady event of the LanguageClient instance. This event is fired when the client is ready and the server has sent back the initializeResult, including the serverInfo.

Here is how you can achieve this:

1. Create and configure the LanguageClient: Set up your language client as usual.
2. Listen for the onReady event: Once the client is ready, you can access the initializeResult and retrieve the serverInfo.version.

Here is an example implementation:

```js
import * as vscode from 'vscode';
import * as path from 'path';
import { LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from 'vscode-languageclient/node';

export function activate(context: vscode.ExtensionContext) {
    // Server configuration
    const serverModule = context.asAbsolutePath(path.join('server', 'out', 'server.js'));
    const debugOptions = { execArgv: ['--nolazy', '--inspect=6009'] };

    const serverOptions: ServerOptions = {
        run: { module: serverModule, transport: TransportKind.ipc },
        debug: { module: serverModule, transport: TransportKind.ipc, options: debugOptions }
    };

    // Client configuration
    const clientOptions: LanguageClientOptions = {
        documentSelector: [{ scheme: 'file', language: 'plaintext' }],
        synchronize: {
            fileEvents: vscode.workspace.createFileSystemWatcher('**/.clientrc')
        }
    };

    // Create the language client
    const client = new LanguageClient(
        'languageServerExample',
        'Language Server Example',
        serverOptions,
        clientOptions
    );

    // Listen for the onReady event
    client.onReady().then(() => {
        const initializeResult = client.initializeResult;
        if (initializeResult && initializeResult.serverInfo) {
            const serverVersion = initializeResult.serverInfo.version;
            console.log(`Server version: ${serverVersion}`);
            // You can now use the server version as needed
        }
    });

    // Start the client
    context.subscriptions.push(client.start());
}

export function deactivate(): Thenable<void> | undefined {
    if (!client) {
        return undefined;
    }
    return client.stop();
}
```

**Reply by dfgordon** [reply]
Hello, this looks like the right idea, but `client.onReady()` was not found for me.  I updated my package.json to use the latest `@types/vscode` and `vscode-languageclient` but still no luck.  Could `onReady` be in the proposed API?

**Comment by dfgordon** [comment]
Thanks Reynald, that lead me to the answer.  Starting with version 8 you can simply `await client.start()`.  Also starting with version 8, `onReady` was removed since `start` returns a promise.  This is mentioned in the changelog [here](https://github.com/Microsoft/vscode-languageserver-node#3170-protocol-800-json-rpc-800-client-and-800-server).
